### ISSUE TITLE
Default usage & recommended best practices are missing from Data Protection docs

### ISSUE THREAD

**Issue by wsloth** [dotnet/AspNetCore.Docs#33135]
### Description

I got the following errors during ASP.NET Core startup:
- `No XML encryptor configured. Key {<guid>} may be persisted to storage in unencrypted form.`
- `Storing keys in a directory '"/root/.aspnet/DataProtection-Keys"' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed. For more information go to https://aka.ms/aspnet/dataprotectionwarning`

When diving further into why this was happening I landed on the linked Data Protection pages. Generally they are very well documented, but what I'm not fully grasping are the following things:
- When is ASP.NET Data Protection used by default?
  - Apparantly it is automatically used (or at least, the logs appear) when using any form of authentication. In my case, I'm using JWT tokens and no cookies, but I'm still getting the warnings even though logically nothing is stored on disk for this use case.
  - I'm also missing in which places it is used by default, for example: CSRF/AntiForgery/TempData/Cookies.
- What is the best practice for configuring it yourself? (KeyVault > Database > Disk storage?)
- When and how should you disable it entirely?
- What can you expect to break when you disable or misconfigure it?

A concrete question/scenario that I'm currently facing:
- I have an API with JWT authentication
- I do not use CSRF/AntiForgery/TempData/Cookies anywhere
- I get the warnings shown above during startup

What should I do in this case?

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/security/data-protection/introduction?view=aspnetcore-8.0

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/security/data-protection/introduction.md

### Document ID

d42b1eab-9682-ea69-f203-e519b456a56f

### Article author

@tdykstra

**Comment by tdykstra** [comment]
@adityamandaleeka can you help with these questions?

**Comment by adityamandaleeka** [comment]
@amcasey

**Comment by amcasey** [comment]
> * When is ASP.NET Data Protection used by default?
>   
>   * Apparantly it is automatically used (or at least, the logs appear) when using any form of authentication. In my case, I'm using JWT tokens and no cookies, but I'm still getting the warnings even though logically nothing is stored on disk for this use case.
>   * I'm also missing in which places it is used by default, for example: CSRF/AntiForgery/TempData/Cookies.

As you've noted, it's relatively difficult to create an aspnetcore app that doesn't pull in Data Protection.  The request to make that possible for JWT-only is here: https://github.com/dotnet/aspnetcore/issues/47410.  Please give it a thumbs-up if it seems to cover your scenario.

> * What is the best practice for configuring it yourself? (KeyVault > Database > Disk storage?)

I don't think there's a one-size-fits-all solution.  If you already have a bunch of Azure resources, then KeyVault is a natural solution.  If you're only running on a single box (or can otherwise share FS access), disk storage might be appropriate (though encrypting that storage may be non-trivial).  Personally, I would probably not add a database to my app just for Data Protection, especially if you're not actually using it.

> * When and how should you disable it entirely?

I think the discussion in https://github.com/dotnet/aspnetcore/issues/47410 is probably relevant.  Please have a look and reply here if you still have questions.

> * What can you expect to break when you disable or misconfigure it?

If you disable it in the sense of not actually protecting anything, you obviously lose protection of data that would otherwise have been stored there.  If you disable it in the sense of removing it from the DI container, you'll probably hit exceptions at runtime.  With a bit of cleverness, you can make a complete-enough shim implementation, but it's rarely a good solution.

> A concrete question/scenario that I'm currently facing:
> 
> * I have an API with JWT authentication
> * I do not use CSRF/AntiForgery/TempData/Cookies anywhere
> * I get the warnings shown above during startup
> 
> What should I do in this case?

Probably just suppress the warning, but see https://github.com/dotnet/aspnetcore/issues/47410.  Happy to discuss further if I've misunderstood your question.

**Comment by nicholasyin** [comment]
This is very annoying, especially the developers don't understand where in our code could generate this warning. If this was caused due to the use of JWT token and Identity, why the warning to developers?

**Comment by wsloth** [comment]
Thank you @amcasey for taking the time for the detailed answer and links to the other issue! I upvoted that one and it seems to be what I'm looking for. 

For the docs: I can only suggest looking into adding in what places data protection is being used so people can better make the estimation of whether they want to fix these warnings or not. For example, not sharing a volume between containers but using sticky sessions in horizontally-scaling set-ups could still work fine, but I found it hard to come to this conclusion myself based off the docs alone.

Feel free to close the issue if no further action will be taken right now
